### DISCUSSION TITLE
Localization

### DISCUSSION THREAD

**Discussion by ec-dev-1** [iTwin/itwinjs-core#3176] [Q&A] [answered]
Good afternoon,

Just wanted to share and discuss an issue I am facing with an iTwin production build in my particular use-case.

We are placing the build files in our repo however, due to the folder structure differences I have to update some of the build folder's static files path declaration otherwise it will not be referenced correctly.

For example,  within  **build  >  static  > CSS** ,  I search up all instance of  "static/media..." and pre-pend  "/apps/itwin" so that in the end, it will be "/apps/itwin/static/media..." 

I also do this for the **static  > JS** so that the MapLayerSources.json and approximateTerrainHeights.json gets referenced correctly during production.

Everything is working as expected however, I cannot get this to work correctly for the **build  > locales** folder.
This results in the variable names being used rather than the actual correct values for example:

See Bottom Bar image:
![Screenshot 2022-02-04 162319](https://user-images.githubusercontent.com/86491141/152610401-43d31136-beed-47a8-abb5-86091f9d6d84.png)

Looking at the network request and searching up .json, you can see that the files within the Locales folder aren't being referenced correctly.

![itwin](https://user-images.githubusercontent.com/86491141/152610421-517b99cb-c830-4959-89fe-77e9aa1e5839.png)

I was able to  update one of the files within the **build > JS** folder so that these json files get referenced correctly.
![Screenshot 2022-02-04 163111](https://user-images.githubusercontent.com/86491141/152610765-52057a2d-2076-4676-8873-542d96f64234.png)

This seems to have a side-effect of creating an unwanted request for   https://localhost:3000/apps/itwin/locales/en-US.json
![Screenshot 2022-02-04 162516](https://user-images.githubusercontent.com/86491141/152610722-46304f87-df5c-4bbc-a783-d72db8a5c15c.png)

Is there a better way to accomplish what I am trying to do?

**Comment by aruniverse** [comment]
I think you want to be setting [IModelAppOptions.publicPath](https://www.itwinjs.org/reference/core-frontend/imodelapp/imodelappoptions/publicpath/). To do this in the itwin viewer you'll need to set the following on the window, like

```ts
(window as any).ITWIN_VIEWER_HOME = `${window.location.origin}/apps/itwin/`
```

You can take a look at https://github.com/iTwin/viewer/blob/master/packages/modules/viewer-react/src/services/BaseInitializer.ts#L222

**Reply by ec-dev-1** [reply]
Thanks @aruniverse .

This definitely reduced the steps I need to take after creating a Production Build as it is now referencing the static files correctly.
I am also getting the confirmation on the console that the resources are being served from the correct path.

I am still facing the issue with the /locales folder.  It is still giving the same issues as before (see image below).
Hovering my mouse of the   _en-US.json_  on the left column tells me that the app is looking for  **apss/itwin/locales/en-US.json** which currently doesn't exist.
 
 Thought I'd report my results while I continue to debug what is going on.
 
 
 Thank you,
 
 
 
![Screenshot 2022-02-07 091847](https://user-images.githubusercontent.com/86491141/152805465-751ebc49-72d9-495a-bd9d-d33b66c31505.png)

**Reply by aruniverse** [reply]
I'm not sure why it says the name of the file it's trying to fetch is `en-US.json`, that definitely doesn't exist, or at least we don't deliver a file with that name.

Do any of the strings in your app show up as localized?

**Reply by ec-dev-1** [reply]
Everything seems to be good except for the following:
- Bottom Bar
- hidden Left Panel
- Sign-out pop-out
- Some tool-tips for built-in tools
![Screenshot 2022-02-07 153916](https://user-images.githubusercontent.com/86491141/152868562-9dbb6390-c0c2-4b7b-897c-532824f471b9.png)
![Screenshot 2022-02-07 153733](https://user-images.githubusercontent.com/86491141/152868564-e829e049-239e-4c7e-8b91-bf89becae86d.png)
![Screenshot 2022-02-04 162319](https://user-images.githubusercontent.com/86491141/152868565-ebb7c06f-f8d9-44c9-8d04-ac6b1713a3ed.png)

**Reply by aruniverse** [reply]
can you open dev tools, run the following in your console and share the output? 

`navigator.languages` (sorry, not a fn)

What version of iTwin.js are you using?
What version of @bentley/react-scripts are you using?

**Comment by ec-dev-1** [comment]
Both Dev and Production appears to be using the same.

![Screenshot 2022-02-07 164443](https://user-images.githubusercontent.com/86491141/152877183-4f90d55a-2811-4153-ba17-3af89f308bee.png)

**Reply by aruniverse** [reply]
What version of iTwin.js are you using?
What version of @bentley/react-scripts are you using?

**Reply by ec-dev-1** [reply]
"@itwin/web-viewer-react": "^1.3.0",
    "@bentley/itwin-client": "^2.19.15",
    "@bentley/react-scripts": "^4.0.3",

Please let me know if you need anymore information.

**Comment by ec-dev-1** [comment, marked as answer]
@aruniverse  upon further investigation, I found that there was a redirect being declared for all locales folder in our repo which affected the iTwin's.  I really appreciate all the assistance you gave while de-bugging.  Thanks!
